modules/camera/camera_x11.cpp

```diff
@@ -0,0 +1,770 @@
+/*  camera_x11.cpp                                                       */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2021 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2021 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "camera_x11.h"
+#include "servers/camera/camera_feed.h"
+
+#include <algorithm>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <dirent.h>
+#include <dlfcn.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <linux/videodev2.h>
+
+#define CLEAR(x) memset(&(x), 0, sizeof(x))
+
+// formats that can be used
+// mainly necessary if libv4l2 is not available
+std::vector<__u32> supported_formats{
+	V4L2_PIX_FMT_RGB24
+};
+
+int V4l2_Device::xioctl(int fd, unsigned long int request, void *arg) {
+	int r;
+	do {
+		r = funcs->ioctl(fd, request, arg);
+	} while (-1 == r && (EINTR == errno || errno == EAGAIN));
+	return r;
```

Comment:
yoda coding! love it but not really our style :)

Reply:
TBH that was just how the example was implemented. I found a little weird too, but hey semantics are the same.^^ I can change this and make it more common if needed.

Reply:
I'll let @akien-mga be the judge of coding styles :)

I do think Yoda coding has merit, it does catch certain errors, but I don't think we've got it anywhere else in the source code.

Reply:
Indeed, we don't use yoda conditions in Godot's codebase as far as I know.

Reply:
So I removed the Yoda conditions to not confuse the codebase.